### PR TITLE
Clarify enumerable cast

### LINKED ISSUE #4142: Enumerable.Cast fail to mention that for value type target it simply unboxes
# Enumerable.Cast fail to mention that for value types it simply unboxes

[Topic: Enumerable.Cast](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.cast?view=netframework-4.7#System_Linq_Enumerable_Cast__1_System_Collections_IEnumerable_).

The following code fails to run and throws an "Invalid cast" exception instead:

    double[] a = new double[10];
    int[] b = a.Cast<int>().ToArray();

There is no mention of this limitation in the documentation and the word "cast" is liberally used which seems to indicate an actual cast operation would be performed. Instead, for value types, a simple unboxing operation is done.

In fact, looking at the code: [Reference Source](https://github.com/Microsoft/referencesource/blob/master/System.Core/System/Linq/Enumerable.cs#L1029):

    foreach (object obj in source) yield return (TResult)obj;

Indicates that every element of the source collection is boxed into an object before being cast (unboxed in this case) to the target type.

**Note**: This issue was sparked by a [question on Stack Overflow: Why does char-array.Cast-to-int raise a cast exception?](https://stackoverflow.com/questions/48317455/why-does-char-castint-raise-a-cast-exception#48317455).

The documentation should be updated to include this limitation and possibly mention *some* parts of the implementation details. If, on the other hand, this method was supposed to do an actual cast, making the first piece of code work then an issue must be filed against the implementation instead.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #4142 

Clarify the allowed conversions for Enumerable.Cast.

/cc @lassevk

### CODE REVIEW

**Comment by BillWagner** [comment]
@rpetrusha This is ready for review. Can you check the VB example I added for correct idiomatic VB?

File: xml/System.Linq/Enumerable.xml [hunk 1]
```diff
@@ -1683,7 +1683,11 @@
   
  The <xref:System.Linq.Enumerable.Cast%60%601%28System.Collections.IEnumerable%29> method enables the standard query operators to be invoked on non-generic collections by supplying the necessary type information. For example, <xref:System.Collections.ArrayList> does not implement <xref:System.Collections.Generic.IEnumerable%601>, but by calling <xref:System.Linq.Enumerable.Cast%60%601%28System.Collections.IEnumerable%29> on the <xref:System.Collections.ArrayList> object, the standard query operators can then be used to query the sequence.  
   
- If an element cannot be cast to type `TResult`, this method will throw an exception. To obtain only those elements that can be cast to type `TResult`, use the <xref:System.Linq.Enumerable.OfType%2A> method instead of <xref:System.Linq.Enumerable.Cast%60%601%28System.Collections.IEnumerable%29>.  
+ If an element cannot be converted to type `TResult`, this method will throw an exception.
```
Comment (rpetrusha) [inline xml/System.Linq/Enumerable.xml]: will throw --> throws
It's best to mention the exception type (InvalidCastException).

File: xml/System.Linq/Enumerable.xml [hunk 2]
```diff
@@ -1683,7 +1683,11 @@
...
+
+ The source sequence for this method is `System.Collections.IEnumerable`, which means the elements have the compile time static type of `object`. The only conversion types that are performed by this method are reference conversions and unboxing conversions. The runtime type of the elements in the collection must match the target type, or in the case of value types, the runtime type of elements must be the result of a boxing conversion of the target type. In practice, other conversion types, such as those between different numeric types are not allowed.
```
Comment (rpetrusha) [inline xml/System.Linq/Enumerable.xml]: Better to do IEnumerable as an xref.
compile time --> compile-time
conversion types --> type conversion
comma after "different numeric types"

File: xml/System.Linq/Enumerable.xml [hunk 3]
```diff
@@ -1694,9 +1698,21 @@ from int i in objects
 ```vb  
 From i As Integer In objects  
 ```  
-  
-   
-  
+
+The `select` clause of a query enables other conversion types, like the implicit numeric conversions. The following example shows using both the `Cast` method and a `select` statement to convert a sequence of boxed integers to a sequence of doubles.
```
Comment (rpetrusha) [inline xml/System.Linq/Enumerable.xml]: shows using both --> uses both

**Review by rpetrusha — APPROVED** [review]
I've left some comments and suggested changes, @BillWagner. You can decide which of them you'd like to address and then merge.

File: xml/System.Linq/Enumerable.xml (hunk 2 above)
Comment (pkulikov) [inline xml/System.Linq/Enumerable.xml]: I've thought a little bit when read about *conversion types* as the sentence before mentions `object` type. So, the word *type* is used in two different meanings in one paragraph. What about:
conversion types --> conversions [or types of conversion]?

And I'd remove *In practice* at the start of the last sentence, as removing it doesn't change a lot, but having it gives impression that in theory things are different.

File: xml/System.Linq/Enumerable.xml (hunk 3 above)
Comment (pkulikov) [inline xml/System.Linq/Enumerable.xml]: Alternative for the first sentence (to be more inline with the style guide):
Use `select` clause of a query to perform other conversion types, like the implicit numeric conversions.

File: xml/System.Linq/Enumerable.xml (hunk 2 above)
Comment (BillWagner) [reply xml/System.Linq/Enumerable.xml]: I think I addressed your first comment via @rpetrusha 's comment above. I like the second.

**Comment by BillWagner** [comment]
build stuck. Closing and reopening.

**Pull request merged**